File: src/calculate_leave.py
from datetime import datetime, timedelta
from typing import Optional
from colorama import Fore
from input_utils import get_user_input
from text_utils import print_cyan, print_green, print_stats


def handle_rush(leave_time: datetime) -> datetime:
    """Adds additional time for rush hour."""

    print_cyan(
        "\nThis order is during rush hour. An additional 15 minutes is suggested."
    )
    additional_minutes = get_user_input("extra rush hour", 15)
    return leave_time - timedelta(minutes=additional_minutes)


def handle_suite(leave_time: datetime, suite_info: str) -> datetime:
    """Adds additional time for a suite."""

    print_cyan(f"\nThis order appears to be for a suite:\n{suite_info}")
    print("\nAn additional 10 minutes is suggested to park and navigate the building.")
    additional_minutes = get_user_input("extra building traversal", 10)
    return leave_time - timedelta(minutes=additional_minutes)


def calculate_leave_time(
    event_time: datetime,
    travel_time: int,
    is_boxes: bool,
    suite_info: Optional[str] = None,
) -> datetime:
    """Calculate the leave time based on the event time, travel time, and whether the order is for a suite."""

    if not isinstance(event_time, datetime):
        raise TypeError("event_time must be a datetime object")

    base_setup = 15 if is_boxes else 30
    print_stats(f"Base travel time: {travel_time} minutes")
    print_stats(
        f"Base setup time: {base_setup} minutes{Fore.CYAN + ' (reduced for box lunch)' if is_boxes else ''}"
    )

    leave_time = event_time - timedelta(minutes=base_setup + travel_time)

    if 16 <= event_time.hour <= 18:
        leave_time = handle_rush(leave_time)

    if suite_info:
        leave_time = handle_suite(leave_time, suite_info)

    leave_time = leave_time.replace(
        minute=leave_time.minute - (leave_time.minute % 5), second=0, microsecond=0
    )

    suggested_minutes = int((event_time - leave_time).total_seconds() / 60)
    print_stats(
        f"\nSuggested setup and travel time: {suggested_minutes} minutes (rounded)."
    )

    user_time = get_user_input("travel and setup", suggested_minutes)
    adjustment = suggested_minutes - user_time
    leave_time += timedelta(minutes=adjustment)

    if adjustment != 0:
        print_green(
            f"Adjusted leave time: Leaving {abs(adjustment)} minutes {'later' if adjustment > 0 else 'earlier'}."
        )

    return leave_time

File: src/leave/calculate_leave.go
package leave

import (
	"fmt"
	"time"

	"github.com/jlsnow301/cutsheet-timer/input"
	"github.com/jlsnow301/cutsheet-timer/utils"
)

// HandleRush adds additional time for rush hour to the leave time.
func handleRush(leaveTime time.Time) time.Time {
	utils.PrintCyan("\nThis order is during rush hour. An additional 15 minutes is suggested.")
	additionalMinutes := input.GetUserInput("extra rush hour", 15)
	return leaveTime.Add(-time.Minute * time.Duration(additionalMinutes))
}

// HandleSuite adds additional time for a suite to the leave time.
func handleSuite(leaveTime time.Time, suiteInfo string) time.Time {
	utils.PrintCyan("\nThis order appears to be for a suite:")
	fmt.Println(suiteInfo)
	fmt.Println("\nAn additional 10 minutes is suggested to park and navigate the building.")
	additionalMinutes := input.GetUserInput("extra building traversal", 10)
	return leaveTime.Add(-time.Minute * time.Duration(additionalMinutes))
}

// CalculateLeaveTime calculates the leave time based on the event time, travel time, and whether the order is for a suite.
func CalculateLeaveTime(eventTime *time.Time, travelTime int, isBoxes bool, suiteInfo string) time.Time {
	baseSetup := 30
	if isBoxes {
		baseSetup = 15
	}

	utils.PrintStats(fmt.Sprintf("Base travel time: %d minutes", travelTime))
	utils.PrintStats(fmt.Sprintf("Base setup time: %d minutes", baseSetup))

	leaveTime := eventTime.Add(-time.Minute * time.Duration(baseSetup+travelTime))

	if eventTime.Hour() >= 16 && eventTime.Hour() <= 18 {
		leaveTime = handleRush(leaveTime)
	}
	if suiteInfo != "" {
		leaveTime = handleSuite(leaveTime, suiteInfo)
	}

	// Round down to the nearest 5 minutes
	leaveTime = leaveTime.Add(-time.Minute * time.Duration(leaveTime.Minute()%5))
	leaveTime = leaveTime.Add(-time.Second * time.Duration(leaveTime.Second()))
	leaveTime = leaveTime.Add(-time.Nanosecond * time.Duration(leaveTime.Nanosecond()))

	suggestedMinutes := int(eventTime.Sub(leaveTime).Minutes())
	utils.PrintStats(fmt.Sprintf("\nSuggested setup and travel time: %d minutes (rounded).", suggestedMinutes))

	userTime := input.GetUserInput("travel and setup", suggestedMinutes)
	adjustment := suggestedMinutes - userTime
	leaveTime = leaveTime.Add(time.Minute * time.Duration(adjustment))

	if adjustment != 0 {
		if adjustment > 0 {
			utils.PrintGreen(fmt.Sprintf("Adjusted leave time: Leaving %d minutes later.", adjustment))
		} else {
			utils.PrintGreen(fmt.Sprintf("Adjusted leave time: Leaving %d minutes earlier.", -adjustment))
		}
	}

	return leaveTime
}
